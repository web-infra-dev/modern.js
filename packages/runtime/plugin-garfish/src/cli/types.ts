import { Config as MasterAppConfig } from '../runtime/useModuleApps';

declare module '@modern-js/app-tools' {
  interface RuntimeUserConfig {
    masterApp?: MasterAppConfig;
  }
}
