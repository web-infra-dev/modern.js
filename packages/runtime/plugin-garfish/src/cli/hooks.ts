import type { Entrypoint } from '@modern-js/types';

export type AppendEntryCodeFn = (params: {
  entrypoint: Entrypoint;
  code: string;
}) => string | Promise<string>;
