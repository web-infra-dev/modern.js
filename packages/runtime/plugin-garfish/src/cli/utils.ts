import type { NormalizedConfig } from '@modern-js/core';

export const makeProvider = () => `
export const provider = function ({basename, dom}) {
  return {
    render({basename, dom, props, appName}) {
      render({ props, basename, dom, appName });
    },
    destroy({ dom }) {
      const node = dom.querySelector('#' + MOUNT_ID) || dom;

      if (node) {
        unmountComponentAtNode(node);
      }
    },
    SubModuleComponent: (props) => {
      const SubApp = render({props, basename});

      return createPortal(<SubApp />, dom.querySelector('#' + MOUNT_ID)  || dom);
    },
    jupiter_submodule_app_key: (props) => {
      const SubApp = render({props, basename});

      return createPortal(<SubApp />, dom.querySelector('#' + MOUNT_ID)  || dom);
    }
  }
};

if (typeof __GARFISH_EXPORTS__ !== 'undefined') {
  __GARFISH_EXPORTS__.provider = provider;
}

function canContinueRender ({ dom, appName }) {
  var renderByGarfish =
    typeof __GARFISH_EXPORTS__ !== 'undefined'
    || typeof window !== 'undefined' && window.Garfish && window.Garfish.activeApps && window.Garfish.activeApps.some((app)=>app.appInfo.name === appName);
  let renderByProvider = dom || appName;
  if (renderByGarfish) {
    // Runs in the Garfish environment and is rendered by the provider
    if (renderByProvider) {
      return true;
    } else {
      // Runs in the Garfish environment and is not rendered by the provider
      return false;
    }
  } else {
    // Running in a non-Garfish environment
    return true;
  }
}

function generateRouterPlugin (basename,routerConfig) {
  if (basename) {
    routerConfig.basename = basename;
    if (routerConfig.supportHtml5History !== false) {
      if (!routerConfig.historyOptions) {
        routerConfig.historyOptions = {
          basename: basename
        };
      } else {
        routerConfig.historyOptions.basename = basename;
      }
    }
  }
  return router(routerConfig);
}

function generateAppWrapperAndRootDom ({ App, props, dom }) {
  let AppWrapper = App;
  if (props) {
    AppWrapper = function () {
      return React.createElement(App, props);
    };
    AppWrapper = hoistNonReactStatics(AppWrapper, App);
  }
  const mountNode = dom ? (dom.querySelector('#' + MOUNT_ID) || dom) : MOUNT_ID;
  return { AppWrapper, mountNode }
}
`;

export const makeRenderFunction = (code: string) => {
  const inGarfishToRender = `
  const { basename, props, dom, appName } = typeof arguments[0] === 'object' && arguments[0] || {};
  if (!canContinueRender({ dom, appName })) return null;
  let { AppWrapper, mountNode } = generateAppWrapperAndRootDom({App, props, dom});
  `;
  return (
    inGarfishToRender +
    code
      .replace(`router(`, `generateRouterPlugin(basename,`)
      .replace('(App)', `(AppWrapper)`)
      .replace('MOUNT_ID', 'mountNode')
  );
};

// support legacy config
export function getRuntimeConfig(config: Partial<NormalizedConfig>) {
  if (config?.runtime?.features) {
    return config?.runtime?.features;
  }
  return config?.runtime || {};
}

// support legacy config
export function setRuntimeConfig(
  config: Partial<NormalizedConfig>,
  key: string,
  value: any,
): undefined {
  if (config?.runtime?.features && config?.runtime?.features[key]) {
    config.runtime.features[key] = value;
    return undefined;
  }
  if (config?.runtime && config?.runtime[key]) {
    config.runtime[key] = value;
    return undefined;
  }
  return undefined;
}
