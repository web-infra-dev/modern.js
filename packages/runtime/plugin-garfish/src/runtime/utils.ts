import { ModuleInfo } from './useModuleApps';

declare const __GARFISH_EXPORTS__: string;

export function isRenderGarfish(params?: { appName?: string }) {
  const renderByGarfish =
    typeof __GARFISH_EXPORTS__ !== 'undefined' ||
    (typeof window !== 'undefined' &&
      window.Garfish &&
      window.Garfish.activeApps &&
      window.Garfish.activeApps.some(
        app => app.appInfo?.name === params?.appName,
      ));
  return renderByGarfish;
}

export const SUBMODULE_APP_COMPONENT_KEY = 'SubModuleComponent';

export function generateSubAppContainerKey(moduleInfo?: ModuleInfo): string {
  return moduleInfo
    ? `modern_sub_app_container_${decodeURIComponent(moduleInfo?.name)}`
    : 'modern_sub_app_container';
}
