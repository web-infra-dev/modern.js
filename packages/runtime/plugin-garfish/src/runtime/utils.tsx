declare const __GARFISH_EXPORTS__: string;

export function isRenderGarfish() {
  return typeof __GARFISH_EXPORTS__ !== 'undefined';
}
