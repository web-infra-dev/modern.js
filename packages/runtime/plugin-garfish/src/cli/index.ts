import path from 'path';
import { createRuntimeExportsUtils, PLUGIN_SCHEMAS } from '@modern-js/utils';
import type {
  CliHookCallbacks,
  CliPlugin,
  useConfigContext,
} from '@modern-js/core';
import { logger } from '../util';
import {
  getRuntimeConfig,
  makeProvider,
  makeRenderFunction,
  setRuntimeConfig,
} from './utils';

export type UseConfig = ReturnType<typeof useConfigContext>;

export const externals = { 'react-dom': 'react-dom', react: 'react' };

export type LifeCycle = CliHookCallbacks;

type NonInValidAble<T> = T extends null | undefined | false ? never : T;

export function getDefaultMicroFrontedConfig(
  microFrontend: NonInValidAble<
    NonNullable<UseConfig['deploy']>['microFrontend']
  >,
) {
  if (microFrontend === true) {
    return {
      enableHtmlEntry: true,
      externalBasicLibrary: false,
      moduleApp: '',
    };
  }

  return {
    enableHtmlEntry: true,
    externalBasicLibrary: false,
    ...microFrontend,
  };
}

export default ({
  runtimePluginName = '@modern-js/runtime/plugins',
  mfPackagePath = path.resolve(__dirname, '../../../../'),
} = {}): CliPlugin => ({
  name: '@modern-js/plugin-garfish',
  setup: ({ useAppContext, useResolvedConfigContext, useConfigContext }) => {
    let pluginsExportsUtils: ReturnType<typeof createRuntimeExportsUtils>;
    let runtimeExportsUtils: ReturnType<typeof createRuntimeExportsUtils>;
    return {
      validateSchema() {
        return PLUGIN_SCHEMAS['@modern-js/plugin-garfish'];
      },
      resolvedConfig: async config => {
        const { resolved } = config;
        const { masterApp, router } = getRuntimeConfig(resolved);
        const nConfig = {
          resolved: {
            ...resolved,
          },
        };
        if (masterApp) {
          // basename does not exist use router's basename
          setRuntimeConfig(
            nConfig.resolved,
            'masterApp',
            Object.assign(
              typeof masterApp === 'object' ? { ...masterApp } : {},
              {
                basename:
                  router?.historyOptions?.basename || router?.basename || '/',
              },
            ),
          );
        }
        logger(`resolvedConfig`, {
          output: nConfig.resolved.output,
          runtime: nConfig.resolved.runtime,
          deploy: nConfig.resolved.deploy,
          server: nConfig.resolved.server,
        });
        return nConfig;
      },
      config() {
        // eslint-disable-next-line react-hooks/rules-of-hooks
        const useConfig = useConfigContext();
        logger('useConfig', useConfig);

        // eslint-disable-next-line react-hooks/rules-of-hooks
        const config = useAppContext();
        pluginsExportsUtils = createRuntimeExportsUtils(
          config.internalDirectory,
          'plugins',
        );
        runtimeExportsUtils = createRuntimeExportsUtils(
          config.internalDirectory,
          'index',
        );

        let disableCssExtract = useConfig.output?.disableCssExtract ?? false;

        // When the micro-frontend application js entry, there is no need to extract css, close cssExtract
        if (useConfig.deploy?.microFrontend) {
          const { enableHtmlEntry } = getDefaultMicroFrontedConfig(
            useConfig.deploy?.microFrontend,
          );
          if (!enableHtmlEntry) {
            disableCssExtract = true;
          }
        }

        return {
          output: {
            disableCssExtract,
          },
          source: {
            alias: {
              '@modern-js/runtime/plugins': pluginsExportsUtils.getPath(),
              '@modern-js/runtime/garfish': mfPackagePath,
            },
          },
          tools: {
            devServer: {
              headers: {
                'Access-Control-Allow-Origin': '*',
              },
            },
            webpackChain: (
              chain,
              {
                webpack,
                env = process.env.NODE_ENV || 'development',
                CHAIN_ID,
              },
            ) => {
              // eslint-disable-next-line react-hooks/rules-of-hooks
              const resolveOptions = useResolvedConfigContext();
              if (resolveOptions?.deploy?.microFrontend) {
                chain.output.libraryTarget('umd');
                if (resolveOptions?.server?.port && env === 'development') {
                  chain.output.publicPath(
                    `//localhost:${resolveOptions.server.port}/`,
                  );
                }

                // add comments avoid sourcemap abnormal
                if (webpack.BannerPlugin) {
                  chain
                    .plugin(CHAIN_ID.PLUGIN.BANNER)
                    .use(webpack.BannerPlugin, [{ banner: 'Micro front-end' }]);
                }

                const { enableHtmlEntry, externalBasicLibrary } =
                  getDefaultMicroFrontedConfig(
                    resolveOptions.deploy?.microFrontend,
                  );
                // external
                if (externalBasicLibrary) {
                  chain.externals(externals);
                }
                // use html mode
                if (!enableHtmlEntry) {
                  chain.output.filename('index.js');
                  chain.plugins.delete(`${CHAIN_ID.PLUGIN.HTML}-main`);
                  chain.optimization.runtimeChunk(false);
                  chain.optimization.splitChunks({
                    chunks: 'async',
                  });
                }
              }
              const resolveWebpackConfig = chain.toConfig();
              logger('webpackConfig', {
                output: resolveWebpackConfig.output,
                externals: resolveWebpackConfig.externals,
                env,
                alias: resolveWebpackConfig.resolve?.alias,
                plugins: resolveWebpackConfig.plugins,
              });
            },
          },
        };
      },
      addRuntimeExports() {
        const addExportStatement = `export { default as garfish, default as masterApp, hoistNonReactStatics } from '${mfPackagePath}'`;
        logger('exportStatement', addExportStatement);
        pluginsExportsUtils.addExport(addExportStatement);
        runtimeExportsUtils.addExport(`export * from '${mfPackagePath}'`);
      },
      modifyEntryImports({ entrypoint, imports }: any) {
        // eslint-disable-next-line react-hooks/rules-of-hooks
        const config = useResolvedConfigContext();
        const { masterApp } = getRuntimeConfig(config);
        if (masterApp) {
          imports.push({
            value: runtimePluginName,
            specifiers: [
              {
                imported: 'garfish',
              },
            ],
          });
          imports.push({
            value: runtimePluginName,
            specifiers: [
              {
                imported: 'masterApp',
              },
            ],
          });
        }
        imports.push({
          value: runtimePluginName,
          specifiers: [
            {
              imported: 'hoistNonReactStatics',
            },
          ],
        });

        imports.push({
          value: 'react-dom',
          specifiers: [
            {
              imported: 'unmountComponentAtNode',
            },
            {
              imported: 'createPortal',
            },
          ],
        });
        return { imports, entrypoint };
      },
      modifyEntryRuntimePlugins({ entrypoint, plugins }: any) {
        // eslint-disable-next-line react-hooks/rules-of-hooks
        const config = useResolvedConfigContext();
        const { masterApp } = getRuntimeConfig(config);
        if (masterApp) {
          logger('garfishPlugin options', masterApp);
          plugins.push({
            name: 'garfish',
            args: 'masterApp',
            options:
              masterApp === true
                ? JSON.stringify({})
                : JSON.stringify(masterApp),
          });
        }
        return { entrypoint, plugins };
      },
      modifyEntryRenderFunction({ entrypoint, code }: any) {
        // eslint-disable-next-line react-hooks/rules-of-hooks
        const config = useResolvedConfigContext();
        if (!config?.deploy?.microFrontend) {
          return { entrypoint, code };
        }
        const nCode = makeRenderFunction(code);
        logger('makeRenderFunction', nCode);
        return {
          entrypoint,
          code: nCode,
        };
      },
      modifyEntryExport({ entrypoint, exportStatement }: any) {
        // eslint-disable-next-line react-hooks/rules-of-hooks
        const config = useResolvedConfigContext();
        if (config?.deploy?.microFrontend) {
          const exportStatementCode = makeProvider();
          logger('exportStatement', exportStatementCode);
          return {
            entrypoint,
            exportStatement: exportStatementCode,
          };
        }
        return {
          entrypoint,
          exportStatement,
        };
      },
    };
  },
});
