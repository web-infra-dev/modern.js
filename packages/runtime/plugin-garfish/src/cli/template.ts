import { formatImportPath } from '@modern-js/utils';

const genRenderCode = ({
  srcDirectory,
  internalSrcAlias,
  metaName,
  entry,
  customEntry,
  customBootstrap,
  mountId,
}: {
  srcDirectory: string;
  internalSrcAlias: string;
  metaName: string;
  entry: string;
  customEntry?: boolean;
  customBootstrap?: string | false;
  mountId?: string;
}) =>
  customEntry
    ? `import '${entry.replace(srcDirectory, internalSrcAlias)}'
export * from '${entry.replace(srcDirectory, internalSrcAlias)}'`
    : `import { createRoot } from '@${metaName}/runtime/react';
import { render } from '@${metaName}/runtime/browser';
import { isRenderGarfish, createProvider } from '@${metaName}/plugin-garfish/tools';
${
  customBootstrap
    ? `import customBootstrap from '${formatImportPath(
        customBootstrap.replace(srcDirectory, internalSrcAlias),
      )}';`
    : 'let customBootstrap;'
}
if (!isRenderGarfish()) {
  const ModernRoot = createRoot();
  ${
    customBootstrap
      ? `customBootstrap(ModernRoot, () => render(<ModernRoot />, '${
          mountId || 'root'
        }'));`
      : `render(<ModernRoot />, '${mountId || 'root'}');`
  };
}

export const provider = createProvider('${
        mountId || 'root'
      }', { customBootstrap });
`;
export const index = ({
  srcDirectory,
  internalSrcAlias,
  metaName,
  entry,
  entryName,
  customEntry,
  customBootstrap,
  mountId,
  appendCode = [],
}: {
  srcDirectory: string;
  internalSrcAlias: string;
  metaName: string;
  entry: string;
  entryName: string;
  customEntry?: boolean;
  customBootstrap?: string | false;
  mountId?: string;
  appendCode?: string[];
}) =>
  `import '@${metaName}/runtime/registry/${entryName}';
  ${genRenderCode({
    srcDirectory,
    internalSrcAlias,
    metaName,
    entry,
    customEntry,
    customBootstrap,
    mountId,
  })}
  ${appendCode.join('\n')}
  `;
