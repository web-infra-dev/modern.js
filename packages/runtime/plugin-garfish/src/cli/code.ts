import path from 'path';
import type {
  AppTools,
  IAppContext,
  NormalizedConfig,
} from '@modern-js/app-tools';
import { fs } from '@modern-js/utils';
import { Entrypoint } from '@modern-js/types';
import type { MaybeAsync } from '@modern-js/core';
import * as template from './template';
import { generateAsyncEntryCode } from './utils';

const ENTRY_POINT_FILE_NAME = 'index.jsx';
export const ENTRY_BOOTSTRAP_FILE_NAME = 'bootstrap.jsx';

export const generateCode = async (
  entrypoints: Entrypoint[],
  appContext: IAppContext,
  config: NormalizedConfig<AppTools>,
  appendEntryCode: (input: { entrypoint: Entrypoint }) => MaybeAsync<string[]>,
) => {
  const { mountId } = config.html;
  const { enableAsyncEntry } = config.source;
  const { internalDirectory, internalSrcAlias, metaName, srcDirectory } =
    appContext;
  await Promise.all(
    entrypoints.map(async entrypoint => {
      const { entryName, isAutoMount, entry, customEntry, customBootstrap } =
        entrypoint;
      const appendCode = await appendEntryCode({ entrypoint });

      if (isAutoMount) {
        // index.jsx
        const indexCode = template.index({
          srcDirectory,
          internalSrcAlias,
          metaName,
          entry,
          entryName,
          customEntry,
          customBootstrap,
          mountId,
          appendCode,
        });
        const indexFile = path.resolve(
          internalDirectory,
          `./${entryName}/${ENTRY_POINT_FILE_NAME}`,
        );

        fs.outputFileSync(indexFile, indexCode, 'utf8');
        if (enableAsyncEntry) {
          const bootstrapFile = path.resolve(
            internalDirectory,
            `./${entryName}/${ENTRY_BOOTSTRAP_FILE_NAME}`,
          );
          fs.outputFileSync(
            bootstrapFile,
            generateAsyncEntryCode(appendCode),
            'utf8',
          );
        }
      }
    }),
  );
};
