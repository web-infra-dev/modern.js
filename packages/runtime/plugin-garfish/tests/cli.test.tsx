import '@testing-library/jest-dom';
import { manager } from '@modern-js/core';
import WebpackChain from 'webpack-chain';
import GarfishPlugin, { externals } from '../src/cli';
import type { UseConfig } from '../src/cli';
import { getRuntimeConfig, makeRenderFunction, setRuntimeConfig } from '../src/cli/utils';
import { CHAIN_ID } from '@modern-js/utils';

describe('plugin-garfish cli', () => {
  test('cli garfish basename', async () => {
    expect(GarfishPlugin().name).toBe('@modern-js/plugin-garfish');

    const main = manager.clone().usePlugin(GarfishPlugin);
    const runner = await main.init();
    await runner.prepare();
    const configHistoryOptions: any = await runner.resolvedConfig({
      resolved: {
        runtime: {
          router: {
            historyOptions: { basename: '/test' }
          },
          masterApp: {}
        },
      }
    } as any);

    expect(configHistoryOptions.resolved.runtime.masterApp.basename).toBe('/test');

    const configHistory: any = await runner.resolvedConfig({
      resolved: {
        runtime: {
          router: {
            basename: '/test2'
          },
          masterApp: {}
        },
      }
    } as any);

    expect(configHistory.resolved.runtime.masterApp.basename).toBe('/test2');
  });

  test('cli get runtime config', ()=>{
    const runtimeConfig = getRuntimeConfig({
      runtime: {
        masterApp: {
          basename: '/test'
        }
      }
    });
    expect(runtimeConfig).toMatchSnapshot();
  });

  test('cli get runtime features config', ()=>{
    const runtimeConfig = getRuntimeConfig({
      runtime: {
        masterApp: {
          basename: '/test'
        },
        features: {
          masterApp: {
            basename: '/test2'
          }
        }
      }
    });

    expect(runtimeConfig).toMatchSnapshot();
  });

  test('cli set runtime config', ()=>{
    const runtimeConfig = {
      runtime: {
        masterApp: {
          basename: '/test'
        }
      }
    };

    setRuntimeConfig(runtimeConfig, 'masterApp', true);

    expect(runtimeConfig.runtime).toMatchSnapshot();
  });

  test('cli set runtime features config', ()=>{
    const runtimeConfig = {
      runtime: {
        features: {
          masterApp: {
            basename: '/test'
          }
        }
      }
    };

    setRuntimeConfig(runtimeConfig, 'masterApp', true);

    expect(runtimeConfig.runtime).toMatchSnapshot();
  });

  test('webpack config close external and use js entry', async ()=>{
    const resolveConfig: any = {
      deploy: {
        microFrontend: {
          externalBasicLibrary: true,
          enableHtmlEntry: false,
        },
      },
      server: {
        port: 8080
      }
    };

    const main = manager.clone({
      useResolvedConfigContext: ()=>resolveConfig
    }).usePlugin(GarfishPlugin);

    const runner = await main.init();
    await runner.prepare();
    const config: any = await runner.config();
    const webpackConfig = new WebpackChain();

    function HTMLWebpackPlugin() {};
    webpackConfig.plugin('html-main').use(HTMLWebpackPlugin);


    config[0].tools.webpackChain(webpackConfig, {
      webpack: jest.fn(),
      env: 'development',
      CHAIN_ID,
    });

    const generateConfig = webpackConfig.toConfig();
    expect(generateConfig).toMatchSnapshot();
    expect(generateConfig).toMatchObject({
      output: {
        libraryTarget: 'umd',
        publicPath: '//localhost:8080/',
        filename: 'index.js'
      },
      externals,
      optimization: { runtimeChunk: false, splitChunks: { chunks: 'async' } }
    });
  })

  test('webpack config default micro config', async ()=>{
    const resolveConfig: any = {
      deploy: {
        microFrontend: true,
      },
      server: {
        port: '8080'
      }
    };

    const main = manager.clone({
      useResolvedConfigContext: ()=>resolveConfig
    }).usePlugin(GarfishPlugin);
    const runner = await main.init();
    await runner.prepare();
    const config: any = await runner.config();
    const webpackConfig = new WebpackChain();
    function HTMLWebpackPlugin() {};
    webpackConfig.plugin('html-main').use(HTMLWebpackPlugin);

    config[0].tools.webpackChain(webpackConfig, {
      webpack: jest.fn(),
      env: 'development'
    });

    const generateConfig = webpackConfig.toConfig();
    expect(config[0].tools.devServer).toMatchObject({
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    });

    expect(generateConfig).toMatchSnapshot();
    expect(generateConfig).toMatchObject({
      output: {
        libraryTarget: 'umd',
        publicPath: '//localhost:8080/'
      }
    });
    expect(generateConfig.externals).toBeUndefined();
    expect(generateConfig.output.filename).toBeUndefined();
  });


  test('micro fronted default config disableCssExtract false', async ()=>{
    const resolveConfig: Partial<UseConfig> = {
      deploy: {
        microFrontend: {},
      }
    };

    const main = manager.clone({
      useResolvedConfigContext: ()=> resolveConfig as any,
      useConfigContext: ()=> resolveConfig,
    }).usePlugin(GarfishPlugin);

    const runner = await main.init();
    await runner.prepare();
    const config = await runner.config();
    expect(config[0].output.disableCssExtract).toBe(false);
  });

  test('micro fronted js entry disableCssExtract true', async ()=>{
    const resolveConfig: Partial<UseConfig> = {
      output: {
        disableCssExtract: false
      },
      deploy: {
        microFrontend: {
          enableHtmlEntry: false,
        },
      }
    };

    const main = manager.clone({
      useResolvedConfigContext: ()=>resolveConfig as any,
      useConfigContext: ()=> resolveConfig,
    }).usePlugin(GarfishPlugin);
    const runner = await main.init();
    await runner.prepare();
    const config = await runner.config();
    expect(config[0].output.disableCssExtract).toBe(true);
  });

  test('normal disableCssExtract false', async ()=>{
    const resolveConfig: Partial<UseConfig> = {};

    const main = manager.clone({
      useResolvedConfigContext: ()=>resolveConfig as any,
      useConfigContext: ()=> resolveConfig,
    }).usePlugin(GarfishPlugin);
    const runner = await main.init();
    await runner.prepare();
    const config = await runner.config();
    expect(config[0].output.disableCssExtract).toBe(false);
  });
});
