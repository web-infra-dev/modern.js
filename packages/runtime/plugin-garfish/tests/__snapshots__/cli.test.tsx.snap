// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin-garfish cli cli get runtime config 1`] = `
Object {
  "masterApp": Object {
    "basename": "/test",
  },
}
`;

exports[`plugin-garfish cli cli get runtime features config 1`] = `
Object {
  "masterApp": Object {
    "basename": "/test2",
  },
}
`;


exports[`plugin-garfish cli cli set runtime config 1`] = `
Object {
  "masterApp": true,
}
`;

exports[`plugin-garfish cli cli set runtime features config 1`] = `
Object {
  "features": Object {
    "masterApp": true,
  },
}
`;

exports[`plugin-garfish cli webpack config close external and use js entry 1`] = `
Object {
  "externals": Object {
    "react": "react",
    "react-dom": "react-dom",
  },
  "optimization": Object {
    "runtimeChunk": false,
    "splitChunks": Object {
      "chunks": "async",
    },
  },
  "output": Object {
    "filename": "index.js",
    "libraryTarget": "umd",
    "publicPath": "//localhost:8080/",
  },
}
`;

exports[`plugin-garfish cli webpack config default micro config 1`] = `
Object {
  "output": Object {
    "libraryTarget": "umd",
    "publicPath": "//localhost:8080/",
  },
  "plugins": Array [
    HTMLWebpackPlugin {},
  ],
}
`;
